### PR TITLE
Added esp_wifi_set_protocol

### PR DESCRIPTION
**Pull request by needlerp**

### CODE REVIEW

File: src/esp32/esp32_wifi.c [hunk 1]
```diff
@@ -327,6 +327,31 @@ bool mgos_wifi_dev_sta_setup(const struct mgos_config_wifi_sta *cfg) {
     tcpip_adapter_dhcpc_start(TCPIP_ADAPTER_IF_STA);
   }
 
+  LOG(LL_INFO, ("Setting STA protocol: %s", cfg->protocol));
+  if (strcmp(cfg->protocol, "B") == 0) {
```
Comment (rojer) [inline src/esp32/esp32_wifi.c]: move this to a common function.
keep style consistent (format manually or run clang-format)

File: mos.yml [hunk 2]
```diff
@@ -28,7 +28,7 @@ config_schema:
   - ["wifi.ap.dhcp_end", "s", "192.168.4.100", {title: "DHCP End Address"}]
   - ["wifi.ap.trigger_on_gpio", "i", -1, {title: "Trigger AP on low GPIO"}]
   - ["wifi.ap.disable_after", "i", 0, {title: "If > 0, will disable itself after the specified number of seconds"}]
-  - ["wifi.ap.hostname", "s", "", {title: "If not empty, DNS server will resolve given host name to the IP address of AP"}]
+  - ["wifi.ap.hostname", "s", "", {title: "If not empty, DNS server will resolve given host name to the IP address of AP"}]
```
Comment (rojer) [inline mos.yml]: remove trailing spaces

File: mos.yml [hunk 3]
```diff
@@ -85,8 +85,10 @@ conds:
       config_schema:
         - ["wifi.ap.keep_enabled", "b", true, {title: "Keep AP enabled when station is on"}]
         - ["wifi.ap.bandwidth_20mhz", "b", false, {title: "enable 20MHz bandwidth AP operation"}]
+        - ["wifi.ap.protocol", "s", "BGN", {title: "Wi-Fi Protocol for AP Mode, defaults to B, G, N"}]
```
Comment (rojer) [inline mos.yml]: describe possible values

File: src/esp32/esp32_wifi.c [hunk 4]
```diff
@@ -327,6 +327,31 @@ bool mgos_wifi_dev_sta_setup(const struct mgos_config_wifi_sta *cfg) {
...
+    r = esp_wifi_set_protocol(WIFI_IF_STA, WIFI_PROTOCOL_11B| WIFI_PROTOCOL_11G|WIFI_PROTOCOL_11N);
+  }
+  else if (strcmp(cfg->protocol, "BGNLR") == 0) {
+    r = esp_wifi_set_protocol(WIFI_IF_STA, WIFI_PROTOCOL_11B| WIFI_PROTOCOL_11G|WIFI_PROTOCOL_11N|WIFI_PROTOCOL_LR);
+  }
+  else if (strcmp(cfg->protocol, "LR") == 0) {
+    r = esp_wifi_set_protocol(WIFI_IF_STA, WIFI_PROTOCOL_LR);
+  }
+  else {
+    r = esp_wifi_set_protocol(WIFI_IF_STA, WIFI_PROTOCOL_11B| WIFI_PROTOCOL_11G|WIFI_PROTOCOL_11N);
```
Comment (rojer) [inline src/esp32/esp32_wifi.c]: it sounds like you can use strchr() to build a bitmask.
```
uin32_t protocol = 0;
if (strchr(cfg->protocol, "B) != NULL) protocol |= WIFI_PROTOCOL_11B;
...
r = esp_wifi_set_protocol(WIFI_IF_STA, protocol);
```

File: mos.yml (hunk 2 above)
Comment (needlerp) [reply mos.yml]: Removed - see updated PR

File: mos.yml (hunk 3 above)
Comment (needlerp) [reply mos.yml]: Possible values added

File: src/esp32/esp32_wifi.c (hunk 1 above)
Comment (needlerp) [reply src/esp32/esp32_wifi.c]: Moved to mgos_wifi_protocol_setup

File: src/esp32/esp32_wifi.c (hunk 4 above)
Comment (needlerp) [reply src/esp32/esp32_wifi.c]: bitmask created, thanks for the tip!

File: include/esp32/esp32_wifi.h [hunk 5]
```diff
@@ -25,6 +25,7 @@ extern "C" {
 #endif
 
 esp_err_t esp32_wifi_ev(system_event_t *event);
+esp_err_t mgos_wifi_protocol_setup(wifi_interface_t ifx, const char *prot);
```
Comment (rojer) [inline include/esp32/esp32_wifi.h]: call it esp32_wifi_protocol_setup

File: src/esp32/esp32_wifi.c [hunk 6]
```diff
@@ -609,3 +622,16 @@ bool mgos_wifi_dev_start_scan(void) {
...
+esp_err_t mgos_wifi_protocol_setup(wifi_interface_t ifx, const char *prot) {
+  uint8_t protocol = 0;
+  esp_err_t r = ESP_OK;
+  LOG(LL_INFO, ("WiFi Setting %d protocol as %s", ifx, prot));
+  if (strchr(prot, 'B') != NULL) protocol |= WIFI_PROTOCOL_11B;
+  if (strchr(prot, 'G') != NULL) protocol |= WIFI_PROTOCOL_11G;
+  if (strchr(prot, 'N') != NULL) protocol |= WIFI_PROTOCOL_11N;
+  if (strpbrk(prot, "LR") != NULL) protocol |= WIFI_PROTOCOL_LR;
+  LOG(LL_INFO, ("WiFi Protocol: %d", protocol));
+  r = esp_wifi_set_protocol(ifx, protocol);
```
Comment (rojer) [inline src/esp32/esp32_wifi.c]: return esp_wifi_set_protocol(ifx, protocol);

File: src/esp32/esp32_wifi.c [hunk 7]
```diff
@@ -609,3 +622,16 @@ bool mgos_wifi_dev_start_scan(void) {
...
+
+esp_err_t mgos_wifi_protocol_setup(wifi_interface_t ifx, const char *prot) {
+  uint8_t protocol = 0;
+  esp_err_t r = ESP_OK;
+  LOG(LL_INFO, ("WiFi Setting %d protocol as %s", ifx, prot));
+  if (strchr(prot, 'B') != NULL) protocol |= WIFI_PROTOCOL_11B;
+  if (strchr(prot, 'G') != NULL) protocol |= WIFI_PROTOCOL_11G;
+  if (strchr(prot, 'N') != NULL) protocol |= WIFI_PROTOCOL_11N;
+  if (strpbrk(prot, "LR") != NULL) protocol |= WIFI_PROTOCOL_LR;
+  LOG(LL_INFO, ("WiFi Protocol: %d", protocol));
```
Comment (rojer) [inline src/esp32/esp32_wifi.c]: combine these two messages into 1, e.g. like so:
```
LOG(LL_INFO, ("WiFi %s: protocol %s (%#x)", (ifx == WIFI_IF_STA ? "STA" : "AP"), prot, protocol));

File: src/esp32/esp32_wifi.c [hunk 8]
```diff
@@ -609,3 +622,16 @@ bool mgos_wifi_dev_start_scan(void) {
...
 }
+
+esp_err_t mgos_wifi_protocol_setup(wifi_interface_t ifx, const char *prot) {
+  uint8_t protocol = 0;
+  esp_err_t r = ESP_OK;
+  LOG(LL_INFO, ("WiFi Setting %d protocol as %s", ifx, prot));
+  if (strchr(prot, 'B') != NULL) protocol |= WIFI_PROTOCOL_11B;
+  if (strchr(prot, 'G') != NULL) protocol |= WIFI_PROTOCOL_11G;
+  if (strchr(prot, 'N') != NULL) protocol |= WIFI_PROTOCOL_11N;
+  if (strpbrk(prot, "LR") != NULL) protocol |= WIFI_PROTOCOL_LR;
```
Comment (rojer) [inline src/esp32/esp32_wifi.c]: strstr(prot, "LR") != NULL

**Pull request merged**